library/eddsa.c

```diff
@@ -0,0 +1,446 @@
+/*
+ *  Edwards-curve Digital Signature Algorithm
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * References:
+ *
+ * SEC1 http://www.secg.org/index.php?action=secg,docs_secg
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_EDDSA_C)
+
+#include "mbedtls/eddsa.h"
+#include "mbedtls/asn1write.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdlib.h>
+#define mbedtls_calloc    calloc
+#define mbedtls_free       free
+#endif
+
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+#if defined(MBEDTLS_ECP_DP_ED25519_ENABLED)
+#include "mbedtls/sha512.h"
+#endif
+
+int mbedtls_eddsa_can_do( mbedtls_ecp_group_id gid )
+{
+    switch( gid )
+    {
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+        case MBEDTLS_ECP_DP_ED25519: return 1;
+#endif
+    default: return 0;
+    }
+}
+
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+static int mbedtls_eddsa_put_dom2_ctx( int flag, const unsigned char *ctx,
+                size_t ctx_len, mbedtls_sha512_context *sha_ctx )
+{
+    unsigned char ct_init_string[] = "SigEd25519 no Ed25519 collisions";
+    unsigned char ct_flag = flag;
+    unsigned char ct_ctx_len = ctx_len & 0xff;
+
+    mbedtls_sha512_update( sha_ctx, ct_init_string, 32 );
+    mbedtls_sha512_update( sha_ctx, &ct_flag, 1 );
+    mbedtls_sha512_update( sha_ctx, &ct_ctx_len, 1 );
+
+    if( ctx && ctx_len > 0)
+        mbedtls_sha512_update( sha_ctx, ctx, ctx_len );
+
+    return( 0 );
+}
+#endif
+
+/*
+ * Compute EdDSA signature of a message.
+ * For PREHASH operation, the message is already previously hashed.
+ * Obviously, for PREHASH, we skip hash message step.
+ */
+int mbedtls_eddsa_sign( mbedtls_ecp_group *grp,
+                mbedtls_mpi *r, mbedtls_mpi *s,
+                const mbedtls_mpi *d, const unsigned char *buf, size_t blen,
+                int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
```

Comment:
It seems that the convention in mbedtls is to pass `f_rng` and `p_rng`last (with the exception for restartable context).